### ISSUE TITLE
Infinite "loop"

### ISSUE THREAD

**Issue by lexterror** [jonelo/jacksum#2]
Hi, 

When I hash anything it goes into an infinite loop hashing the same files infinitely!

However, when I click on the command prompt it stops!

**Comment by jonelo** [comment]
> However, when I click on the command prompt it stops!

Well, actually that is normal behavior on the Windows cmd. If you click in the Window of the command prompt it pauses the output. If you hit Enter in the Window of the command prompt it should continue with the output again.

Thanks lexterror for creating a new ticket! I copy the relevant line from ticket #1:
```
jacksum -a sha3-512 -f -F "#FINGERPRINT #FILESIZE #FILENAME" *
```
The star at the end will be expanded by your shell. Files in the current folder are being hashed, and if there are folders in your current working directory, those folders will be taken into account as well and each folder will be traversed recursively with max. depth by default. Dependent on the number of files in the folders, it could take some time. Also Jacksum 3 uses all processors/cores to process the files simultanously, and since some files are larger than others, larger files could be printed later and the order where lines appear are not necessarily in order.

Does that lead to the impression of a loop?

The question is whether you really see lines appearing more than once in the output? 

In order to track down the issue further, I suggest to save the output to a file to verify that there are duplicate lines. You can use OS redirection or you can use the options -O, -o, and -u, and -U.
You can also control the behavior by limiting the walking depth by setting -r.
Also option -d could be of interest for you:
```
jacksum -h -d
    -d
    --dont-follow-symlinks-to-directories
            Don't follow symbolic links on directories, that means regular
            directories only. Starting with version 3, Jacksum detects file
            system cycles (a symbolik link from a sub folder to a parent folder)
            and skips folders that could cause an endless loop while a
            recursively traversal. If you set -d, Jacksum doesn't follow any
            symbolik links on directories.
            Available since Jacksum 1.5.0, see also -r, -f
            File system cycle detection available since Jacksum 3.0.0
```
Does it stop the "loop" if you set -d ?
If you just want to hash the files in the current working directory and no folders at all, you could replace the star by a dot (which represents the current working directory) and limit the depth to 1 by seeting option -r:
```
jacksum -a sha3-512 -f -F "#FINGERPRINT #FILESIZE #FILENAME" -r 1 .
```
Please let me know if you really see duplicate lines in the output, an output file should help to verify it.
Fingers crossed ;-)

**Comment by lexterror** [comment]
The problem is still there! I'm attaching a screenshot! I also tried both -d and -r 1
![desktop](https://user-images.githubusercontent.com/16135535/132144374-185ffa3a-63d6-462e-a364-f0f7ea6ecd02.png)

**Comment by lexterror** [comment]
I just tried both "*" and "." (it's the same output)

**Comment by jonelo** [comment]
The line starting with "Jacksum: Warning: Option -r has been set, ..." appears on the screenshot multiple times actually. However, that warning is part of the initialization function of Jacksum and will be executed only once before it reads any file. Since we see that line multiple times on the screenshot, it means Jacksum has been launched multiple times - each one is a separate process actually!

Since you are on Windows ... are you calling Jacksum from a batch file?
How does that batch look like?

**Comment by lexterror** [comment]
Yes, I'm using a batch file!

@echo off
:: ##############################################################################
:: # Use this batch file for launching Jacksum easily from the command line
:: # just by typing "jacksum". You can remove this comment block if you like.
:: #
:: # You need to edit the path below (C:\Program Files ...) so jacksum.jar
:: # can be found!
:: # Make also sure, that this batch file is reachable by the
:: # operating system environment variable called PATH.
:: #
:: # The %* modifier is a unique modifier that represents all arguments
:: # passed in a batch file (e. g. %1 %2 %3 %4 %5 ...%255), but it may
:: # not work on all Windows platforms. In this case just use the old syntax:
:: # java -jar "C:\Program Files\Jacksum\jacksum.jar" %1 %2 %3 %4 %5 %6 %7 %8 %9
:: ##############################################################################

java -jar "jacksum.jar" %*

jacksum -a sha3-512 -f -F "#FINGERPRINT #FILESIZE #FILENAME" -r 1 *

**Comment by jonelo** [comment]
Your batch file is called jacksum.bat. The last line: if you call "jacksum", it calls jacksum.bat actually, and jacksum.bat calls jacksum.bat, and so on. It is a recursion without end condition, the batch calls itself forever!

In order to fix that issue you can replace the line
```
java -jar "jacksum.jar" %*
```
with
```
java -jar "jacksum.jar" -a sha3-512 -f -F "#FINGERPRINT #FILESIZE #FILENAME" %*
```
and then you can call
```
jacksum -r 1 .
```
or 
```
jacksum *
```
for example.

Does that help?

**Comment by lexterror** [comment]
Solved! Thank you!
